src/monitor.ts

```diff
@@ -0,0 +1,357 @@
+  }
+
+  return data;
+};
+
+/**
+ * An internal interface representing the state we keep track of for each
+ * network node.
+ */
+interface IMonitoredNodeStatus {
+  address: string;
+  alive: boolean;
+  lastUpdate: Date;
+  lastResponseId: number;
+  networkId?: string;
+  reviveCmd?: string;
+  reviveArgs?: string[];
+};
+
+/**
+ * An internal interface representing the response of the ping command to the
+ * network node.
+ */
+interface INetVersionResponse {
+  id: number;
+  result: string;
+};
+
+/**
+ * The actual monitor implementation. It contains an express application and
+ * last known node status.
+ */
+class Monitor {
+  /**
+   * The interval at which the monitor pings the instances. If the last update
+   * time was longer than this interval, the instance is considered dead.
+   */
+  static k_HEARTBEAT_INTERVAL = 5000;
+
+  /**
+   * The web server
+   */
+  server: express.Application;
+
+  /**
+   * The current state of the nodes
+   */
+  nodesStatus: IMonitoredNodeStatus[];
+
+  /**
+   * The timer used to ping nodes
+   */
+  timer: NodeJS.Timer;
+
+  private validateNetVersionResponse(data: any) : INetVersionResponse | undefined {
+    if (!data.hasOwnProperty('id')) {
+      return undefined;
+    }
+
+    if (typeof data.id !== 'number') {
+      return undefined;
+    }
+
+    if (!data.hasOwnProperty('result')) {
+      return undefined;
+    }
+
+    return <INetVersionResponse>data;
+  }
+
+  private configToInitialNodeStatus(config: IMonitoredNodeConfig) : IMonitoredNodeStatus {
+    var result : IMonitoredNodeStatus = {
+      address: config.address,
+      alive: false,
+      lastUpdate: new Date(),
+      lastResponseId: 0
+    };
+
+    if (typeof config.reviveCmd !== 'undefined') {
+      result.reviveCmd = config.reviveCmd;
+    }
+
+    if (typeof config.reviveArgs !== 'undefined') {
+      result.reviveArgs = config.reviveArgs.split(/\s+/);
+    }
+
+    return result;
+  }
+
+  private setupRouting() {
+    // GET /status returns the current status of all known instances.
+    this.server.get('/status', (req, res) => {
+      res.send(this.nodesStatus);
+    });
+
+    // Parse all request body as JSON, ignoring content-type, for the /add
+    // route.
+    this.server.use('/add', bodyParser.json({
+      type: '*/*'
+    }));
+
+    // POST /add adds the provided instances to the monitored list. The POST
+    // data must be a JSON object with the following format.
+    // {
+    //   nodes: [IMonitoredNodeConfig...]
+    // }
+    this.server.post('/add', (req, res) => {
+
+      winston.info(req.body);
+
+      if (!req.body.hasOwnProperty('nodes')) {
+        res.status(404).send('There is no field named nodes.');
+        return;
+      }
+
+      if (!(req.body.nodes instanceof Array)) {
+        res.status(404).send('The field nodes is not an array.');
+        return;
+      }
+
+      for (let data of req.body.nodes) {
+        const monitoredNodeConfig = validateMonitoredNodeConfig(data);
```

Comment:
i wonder if something like [express-validator](https://github.com/ctavan/express-validator) can be employed here